wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -1418,25 +1427,90 @@ class KotlinGenerator private constructor(
     else -> nameToKotlinName.getValue(this)
   }
 
-  private fun Field.getClass(baseClass: TypeName = type!!.asTypeName()) = when {
-    isRepeated -> List::class.asClassName().parameterizedBy(baseClass)
-    isOptional -> baseClass.copy(nullable = true)
-    else -> baseClass.copy(nullable = false)
+  private fun Field.getClass(baseClass: TypeName = type!!.asTypeName()): TypeName {
+    return when (encodeMode!!) {
+      EncodeMode.REPEATED,
+      EncodeMode.PACKED -> List::class.asClassName().parameterizedBy(baseClass)
+      EncodeMode.NULL_IF_ABSENT -> {
+        if (isOneOf || isMap) baseClass.copy(nullable = false)
+        else baseClass.copy(nullable = true)
+      }
+      else -> baseClass.copy(nullable = false)
+    }
   }
 
   private val Field.typeName: TypeName
-    get() = when {
-      isRepeated -> List::class.asClassName().parameterizedBy(type!!.typeName)
-      isMap -> Map::class.asTypeName().parameterizedBy(keyType.typeName, valueType.typeName)
-      !isRequired -> type!!.typeName.copy(nullable = true)
-      else -> type!!.typeName
+    get() {
+      if (isMap) {
+        return Map::class.asTypeName().parameterizedBy(keyType.typeName, valueType.typeName)
+      }
+
+      return when (encodeMode!!) {
+        EncodeMode.REPEATED,
+        EncodeMode.PACKED -> List::class.asClassName().parameterizedBy(type!!.typeName)
+        EncodeMode.NULL_IF_ABSENT -> type!!.typeName.copy(nullable = true)
+        EncodeMode.THROW_IF_ABSENT -> type!!.typeName
+        EncodeMode.IDENTITY_IF_ABSENT -> {
+          when {
+            isOneOf -> type!!.typeName.copy(nullable = true)
+            type!!.isMessage -> type!!.typeName.copy(nullable = true)
+            else -> type!!.typeName
+          }
+        }
+      }
     }
 
   private val Field.defaultValue: CodeBlock
-    get() = when {
-      isRepeated -> CodeBlock.of("emptyList()")
-      isMap -> CodeBlock.of("emptyMap()")
-      else -> CodeBlock.of("null")
+    get() {
+      if (isMap) return CodeBlock.of("emptyMap()")
+
+      return when (encodeMode!!) {
+        EncodeMode.REPEATED,
+        EncodeMode.PACKED -> CodeBlock.of("emptyList()")
+        EncodeMode.NULL_IF_ABSENT -> CodeBlock.of("null")
```

Comment:
this can be true for proto2 ? In which case there could be a different default value?

Reply:
```
optional CurrencyCode  currency = 1 [default = USD]
```

Reply:
I don’t like using the word “default” ’cause that’s taken

Reply:
renamed the method to `endentityValue`